### PR TITLE
Store auth in keyring and transfer it to nydusd via environment variable

### PR DESCRIPTION
**Pull request by sctb512**
1. Store `auth` for `nydusd` in the keyring instead of the configuration file.
2. Set `auth` via environment variable to avoid displaying it in command parameter.

Related to https://github.com/dragonflyoss/image-service/pull/1381 and https://github.com/dragonflyoss/image-service/pull/1382

### CODE REVIEW

File: pkg/manager/daemon_adaptor.go [hunk 1]
```diff
@@ -192,6 +195,19 @@ func (m *Manager) BuildDaemonCommand(d *daemon.Daemon, bin string, upgrade bool)
 
 	cmd := exec.Command(nydusdPath, args...)
 
+	if config.IsKeyringEnabled() && !config.IsFusedevSharedModeEnabled() {
```
Comment (changweige) [inline pkg/manager/daemon_adaptor.go:198]: Daemon has its own working mode which means nydusd daemons can work in the different modes in a single node.
Better to judge by daemon, we already have some helper functions doing this

File: pkg/filesystem/fs.go [hunk 2]
```diff
@@ -298,7 +300,14 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string) (err er
 			daemonconfig.CacheDir:  cacheDir,
 		}
 		cfg := deepcopy.Copy(fsManager.DaemonConfig).(daemonconfig.DaemonConfig)
-		err = daemonconfig.SupplementDaemonConfig(cfg, imageID, snapshotID, false, labels, params)
+		var updateErr error
+		err = daemonconfig.SupplementDaemonConfig(cfg, imageID, snapshotID, false, labels, params, func(imageHost string, keyChain *auth.PassKeyChain) {
+			logrus.Infof("add key for %s", imageHost)
```
Comment (changweige) [inline pkg/filesystem/fs.go:305]: It's better to lower the log level to Debug

File: pkg/manager/daemon_adaptor.go [hunk 3]
```diff
@@ -192,6 +195,19 @@ func (m *Manager) BuildDaemonCommand(d *daemon.Daemon, bin string, upgrade bool)
...
+		c, err := daemonconfig.NewDaemonConfig(d.States.FsDriver, configPath)
```
Comment (changweige) [inline pkg/manager/daemon_adaptor.go:199]: I am afraid the current way to identify a container image reference's host is not agile.  For the dedicated mode nydusd daemon, it only as a rafs instance in its Instances where we can calculate what host is

File: pkg/keyring/keyring.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package keyring
```
Comment (changweige) [inline pkg/keyring/keyring.go:1]: Please add license claim

File: pkg/daemon/daemon.go [hunk 5]
```diff
@@ -66,6 +71,7 @@ type Daemon struct {
 	// fusedev shared mode: zero, one or more RAFS instances
 	// fscache shared mode: zero, one or more RAFS instances
 	Instances rafsSet
+	authCache *lru.Cache
```
Comment (changweige) [inline pkg/daemon/daemon.go:74]: Auth should be shared among different nydusd daemons. So why to bind the cache to a single nydusd daemon?

File: pkg/daemon/daemon.go [hunk 6]
```diff
@@ -151,6 +158,31 @@ func (d *Daemon) RemoveInstance(snapshotID string) {
 	d.DecRef()
 }
 
+func (d *Daemon) UpdateAuth(imageHost, auth string) error {
+	key, err := keyring.Add(imageHost, auth)
```
Comment (changweige) [inline pkg/daemon/daemon.go:162]: Is it necessary to consider the atomicity when updating?

File: pkg/daemon/daemon.go [hunk 7]
```diff
@@ -151,6 +158,31 @@ func (d *Daemon) RemoveInstance(snapshotID string) {
...
+	return nil
+}
+
+func (d *Daemon) GetAuth(imageHost string) (string, error) {
+	if auth, ok := d.authCache.Get(imageHost); ok {
+		return auth.(string), nil
+	}
+
+	data, err := keyring.Search(imageHost)
+	if err == nil {
```
Comment (changweige) [inline pkg/daemon/daemon.go:180]: Checking if err is nil looks strange.

**Comment by imeoer** [comment]
This approach seems to only solve the nydusd registry credential persistence problem, but when the credential is changed (e.g. registry user/pass dynamically generated by a credential helper), nydusd may not be able to timely detect the update. I think https://github.com/containerd/nydus-snapshotter/pull/504 is a better implementation.

File: pkg/manager/daemon_adaptor.go (hunk 1 above)
Comment (sctb512) [reply pkg/manager/daemon_adaptor.go]: Fixed.

File: pkg/filesystem/fs.go (hunk 2 above)
Comment (sctb512) [reply pkg/filesystem/fs.go]: Fixed.

File: pkg/manager/daemon_adaptor.go (hunk 3 above)
Comment (sctb512) [reply pkg/manager/daemon_adaptor.go]: Fixed.

File: pkg/keyring/keyring.go (hunk 4 above)
Comment (sctb512) [reply pkg/keyring/keyring.go]: Fixed.

File: pkg/daemon/daemon.go (hunk 5 above)
Comment (sctb512) [reply pkg/daemon/daemon.go]: Fixed. Move it to the manager.

File: pkg/daemon/daemon.go (hunk 6 above)
Comment (sctb512) [reply pkg/daemon/daemon.go]: Done.

File: pkg/daemon/daemon.go (hunk 7 above)
Comment (sctb512) [reply pkg/daemon/daemon.go]: Fixed.

**Comment by sctb512** [comment]
> This approach seems to only solve the nydusd registry credential persistence problem, but when the credential is changed (e.g. registry user/pass dynamically generated by a credential helper), nydusd may not be able to timely detect the update. I think #504 is a better implementation.

I believe this PR does not conflict with https://github.com/containerd/nydus-snapshotter/pull/504. It is better to get authentication from the credential server if `snapshotter` does not send it through `env`.

File: pkg/auth/cache.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package auth
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/utils/lru"
```
Comment (imeoer) [inline pkg/auth/cache.go:11]: It seems we have used `github.com/golang/groupcache/lru`.

File: pkg/auth/keyring.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+const (
+	defaultSessionName = "_ses.nydus"
+)
+
+type KeyRing struct {
+	lock *sync.RWMutex
+}
+
+func NewKeyRing() *KeyRing {
```
Comment (imeoer) [inline pkg/auth/keyring.go:26]: Maybe give some unit test cases for this?

File: pkg/manager/daemon_adaptor.go [hunk 10]
```diff
@@ -192,6 +199,25 @@ func (m *Manager) BuildDaemonCommand(d *daemon.Daemon, bin string, upgrade bool)
 
 	cmd := exec.Command(nydusdPath, args...)
 
+	if config.IsKeyringEnabled() && !d.IsSharedDaemon() {
```
Comment (imeoer) [inline pkg/manager/daemon_adaptor.go:202]: Can we support shared daemon?

File: config/daemonconfig/daemonconfig.go [hunk 11]
```diff
@@ -33,6 +33,7 @@ type DaemonConfig interface {
 	FillAuth(kc *auth.PassKeyChain)
 	StorageBackend() (StorageBackendType, *BackendConfig)
 	UpdateMirrors(mirrorsConfigDir, registryHost string) error
+	GetImageHost() string
```
Comment (changweige) [inline config/daemonconfig/daemonconfig.go:36]: Seems there is no caller for the added method `GetImageHost`. Please remove it if so

File: config/daemonconfig/daemonconfig.go [hunk 12]
```diff
@@ -168,7 +169,11 @@ func SupplementDaemonConfig(c DaemonConfig, imageID, snapshotID string,
 		// when repository is public.
 		keyChain := auth.GetRegistryKeyChain(registryHost, imageID, labels)
 		c.Supplement(registryHost, image.Repo, snapshotID, params)
-		c.FillAuth(keyChain)
+		if config.IsKeyringEnabled() && fn != nil {
+			fn(registryHost, keyChain)
```
Comment (changweige) [inline config/daemonconfig/daemonconfig.go:173]: Can `fn` fail ?

File: pkg/daemon/daemon.go [hunk 13]
```diff
@@ -232,13 +236,13 @@ func (d *Daemon) SharedMount(rafs *Rafs) error {
...
 	case config.FsDriverFusedev:
-		return d.sharedFusedevMount(rafs)
+		return d.sharedFusedevMount(rafs, authCache)
 	default:
 		return errors.Errorf("unsupported fs driver %s", d.States.FsDriver)
 	}
 }
 
-func (d *Daemon) sharedFusedevMount(rafs *Rafs) error {
+func (d *Daemon) sharedFusedevMount(rafs *Rafs, authCache *auth.Cache) error {
```
Comment (changweige) [inline pkg/daemon/daemon.go:245]: Erofs/fscache fs driver should also need the auth

File: pkg/daemon/daemon.go [hunk 14]
```diff
@@ -255,6 +259,25 @@ func (d *Daemon) sharedFusedevMount(rafs *Rafs) error {
 			d.ConfigFile(rafs.SnapshotID))
 	}
 
+	if config.IsKeyringEnabled() {
+		image, err := registry.ParseImage(rafs.ImageID)
```
Comment (changweige) [inline pkg/daemon/daemon.go:263]: It's would be great if we can wrap this code snippet into a helper function which tries to fill the auth into config string

File: pkg/filesystem/fs.go [hunk 15]
```diff
@@ -298,7 +300,14 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string) (err er
 			daemonconfig.CacheDir:  cacheDir,
 		}
 		cfg := deepcopy.Copy(fsManager.DaemonConfig).(daemonconfig.DaemonConfig)
-		err = daemonconfig.SupplementDaemonConfig(cfg, imageID, snapshotID, false, labels, params)
+		var updateErr error
+		err = daemonconfig.SupplementDaemonConfig(cfg, imageID, snapshotID, false, labels, params, func(imageHost string, keyChain *auth.PassKeyChain) {
+			logrus.Debugf("add key for %s", imageHost)
```
Comment (changweige) [inline pkg/filesystem/fs.go:305]: We don't usually directly use logrus in this project but using log.L

File: pkg/filesystem/fs.go [hunk 16]
```diff
@@ -298,7 +300,14 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string) (err er
...
+			updateErr = fsManager.AuthCache.UpdateAuth(imageHost, keyChain.ToBase64())
+		})
+		if updateErr != nil {
```
Comment (changweige) [inline pkg/filesystem/fs.go:308]: Passing closure's error code through an intermediate variable is a little tricky. How about give the closure a return value which can be returned by `SupplementDaemonConfig`

File: pkg/auth/keyring.go (hunk 9 above)
Comment (changweige) [reply pkg/auth/keyring.go:26]: Agree, please try to add some unit test, especially for keying garbage collection and if the boundary is correctly handled.

File: pkg/auth/keyring.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	return &KeyRing{}
+}
+
+func (k *KeyRing) Add(id, value string) (int, error) {
+	sessKeyID, err := unix.KeyctlJoinSessionKeyring(defaultSessionName)
+	if err != nil {
+		return 0, errors.Wrap(err, "unable to session key")
+	}
+
+	if err := k.modifyKeyringPerm(sessKeyID, 0xffffffff, 0x80000); err != nil {
```
Comment (changweige) [inline pkg/auth/keyring.go:36]: Please add a comment telling what those magic numbers for

File: pkg/auth/keyring.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func NewKeyRing() *KeyRing {
+	return &KeyRing{}
+}
+
+func (k *KeyRing) Add(id, value string) (int, error) {
+	sessKeyID, err := unix.KeyctlJoinSessionKeyring(defaultSessionName)
+	if err != nil {
+		return 0, errors.Wrap(err, "unable to session key")
```
Comment (changweige) [inline pkg/auth/keyring.go:33]: join the session key

File: pkg/auth/keyring.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	if err := k.modifyKeyringPerm(sessKeyID, 0xffffffff, 0x80000); err != nil {
+		return 0, errors.Wrap(err, "unable to mod keyring permissions")
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
```
Comment (changweige) [inline pkg/auth/keyring.go:41]: Why do we need a lock here? Is `unix.AddKey` not atomic?

File: pkg/auth/keyring.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return err
+	}
+	perm := (uint32(perm64) & mask) | setbits
+
+	return unix.KeyctlSetperm(ringID, perm)
+}
+
+func (k *KeyRing) Search(id string) (string, error) {
+	sessKeyID, err := unix.KeyctlJoinSessionKeyring(defaultSessionName)
```
Comment (changweige) [inline pkg/auth/keyring.go:72]: Is it necessary to join the keyring session again and again?

File: pkg/auth/keyring.go [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return "", errors.Wrap(err, "unable to session key")
+	}
+
+	if err := k.modifyKeyringPerm(sessKeyID, 0xffffffff, 0x80000); err != nil {
```
Comment (changweige) [inline pkg/auth/keyring.go:77]: Don't use negative sentence int he wrapped error please

File: pkg/auth/keyring.go [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		return "", errors.Wrap(err, "unable to mod keyring permissions")
+	}
+
+	key, err := unix.KeyctlSearch(sessKeyID, "user", id, 0)
+	if err != nil {
+		return "", err
```
Comment (changweige) [inline pkg/auth/keyring.go:83]: Please wrap the error

File: pkg/auth/keyring.go [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+	sizeRead := size + 1
+
+	for sizeRead > size {
+		sizeRead, err = unix.KeyctlBuffer(unix.KEYCTL_READ, key, buffer, size)
+		if err != nil {
+			return "", err
+		}
+
+		if sizeRead > size {
+			buffer = make([]byte, sizeRead)
```
Comment (changweige) [inline pkg/auth/keyring.go:102]: Um. How is it possible `sizeRead` is greater then `size`?

Comment (changweige) [reply pkg/auth/keyring.go:102]: Please try to add a UT for it

File: pkg/auth/keyring.go [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+	return keyID, nil
+}
+
+func (k *KeyRing) modifyKeyringPerm(ringID int, mask, setbits uint32) error {
+	dest, err := unix.KeyctlString(unix.KEYCTL_DESCRIBE, ringID)
+	if err != nil {
+		return err
+	}
+
+	res := strings.Split(dest, ";")
```
Comment (changweige) [inline pkg/auth/keyring.go:57]: It's better to add a output example as comment to helper code readers to understand

File: pkg/auth/keyring.go (hunk 19 above)
Comment (sctb512) [reply pkg/auth/keyring.go:41]: We need this. `Addkey` only launch a `syscall`, but can't make sure it's atomic.
```shell
r0, _, e1 := Syscall6(SYS_ADD_KEY, uintptr(unsafe.Pointer(_p0)), uintptr(unsafe.Pointer(_p1)), uintptr(_p2), uintptr(len(payload)), uintptr(ringid), 0)
```

File: pkg/manager/daemon_adaptor.go (hunk 10 above)
Comment (sctb512) [reply pkg/manager/daemon_adaptor.go:201]: Yes, it call `c.FillAuth(keyChain)` to fill auth before sending mount request to nydus daemon.

File: config/daemonconfig/daemonconfig.go (hunk 11 above)
Comment (sctb512) [reply config/daemonconfig/daemonconfig.go]: Done.

File: config/daemonconfig/daemonconfig.go (hunk 12 above)
Comment (sctb512) [reply config/daemonconfig/daemonconfig.go]: Yes, it depends on the caller `SupplementDaemonConfig` and the `fn` may fail.

File: pkg/daemon/daemon.go (hunk 13 above)
Comment (sctb512) [reply pkg/daemon/daemon.go:244]: Fixed.

File: pkg/daemon/daemon.go (hunk 14 above)
Comment (sctb512) [reply pkg/daemon/daemon.go]: Do you mean to wrap the code for parsing images and get the `keyChain` to a function?

Comment (sctb512) [reply pkg/daemon/daemon.go]: > Do you mean to wrap the code for parsing images and get the `keyChain` to a function?

Done.

File: pkg/filesystem/fs.go (hunk 15 above)
Comment (sctb512) [reply pkg/filesystem/fs.go]: Fixed.

File: pkg/filesystem/fs.go (hunk 16 above)
Comment (sctb512) [reply pkg/filesystem/fs.go]: Done.

File: pkg/auth/keyring.go (hunk 17 above)
Comment (sctb512) [reply pkg/auth/keyring.go]: Done.

File: pkg/auth/keyring.go (hunk 18 above)
Comment (sctb512) [reply pkg/auth/keyring.go]: Fixed.

File: pkg/auth/keyring.go (hunk 21 above)
Comment (sctb512) [reply pkg/auth/keyring.go]: Fixed.

File: pkg/auth/keyring.go (hunk 22 above)
Comment (sctb512) [reply pkg/auth/keyring.go]: DOne.

File: pkg/auth/keyring.go (hunk 23 above)
Comment (sctb512) [reply pkg/auth/keyring.go]: Done.

File: pkg/auth/keyring.go (hunk 24 above)
Comment (sctb512) [reply pkg/auth/keyring.go:116]: Done.

File: config/daemonconfig/daemonconfig.go [hunk 25]
```diff
@@ -168,7 +169,16 @@ func SupplementDaemonConfig(c DaemonConfig, imageID, snapshotID string,
 		// when repository is public.
 		keyChain := auth.GetRegistryKeyChain(registryHost, imageID, labels)
 		c.Supplement(registryHost, image.Repo, snapshotID, params)
-		c.FillAuth(keyChain)
+		if config.IsKeyringEnabled() && fn != nil {
+			if err := fn(registryHost, keyChain); err != nil {
+				if errors.Is(err, unix.EINVAL) {
```
Comment (changweige) [inline config/daemonconfig/daemonconfig.go:174]: I think the error type should be transparent for the caller

**Review by changweige — COMMENTED** [review]
Could we also have an e2e test case defined in `integration/entrypoint.sh`
